mlflow/proxy/server.py

```diff
@@ -0,0 +1,171 @@
+import argparse
+import signal
+from datetime import datetime, timedelta
+from fastapi import FastAPI, Request, HTTPException
+from fastapi.responses import JSONResponse
+import os
+import sys
+from typing import Dict, Callable
+
+from mlflow.proxy.proxy_service_constants import _Route
+from mlflow.proxy.utils import (
+    _error_response,
+    _load_route_config,
+    _get_route_limits,
+    _parse_request_path,
+    _parse_url_path_for_base_url,
+)
+
+app = FastAPI()
+
+active_routes: Dict[str, _Route] = {}
+
+config_path = os.environ["MLFLOW_PROXY_CONFIG_PATH"]
+configs = _load_route_config(config_path)
+route_limits = _get_route_limits(configs)
+
+request_counts: Dict[str, int] = {}
+first_requests: Dict[str, datetime] = {}
+
+TIME_WINDOW = timedelta(minutes=1)
+
+PID = os.getpid()
+
+
+@app.get("/health")
+def health():
+    return {"status": "OK"}
+
+
+@app.get("/route_statuses")
+async def get_route_statuses():
+    return active_routes
+
+
+@app.get("/disable_route/{route:path}")
+async def disable_route(route: str):
+    if route in active_routes:
+        if active_routes[route].active:
+            active_routes[route].active = False
+            return {"status": f"Route {route} disabled"}
+        else:
+            return _error_response(405, f"Route {route} is already disabled")
+    else:
+        return _error_response(404, f"There is no route {route} activated on this server")
+
+
+@app.get("/enable_route/{route:path}")
+async def enable_route(route: str):
+    if route in active_routes:
+        if not active_routes[route].active:
+            active_routes[route].active = True
+            return {"status": f"Route {route} enabled"}
+        else:
+            _error_response(405, f"Route {route} is already enabled")
+    else:
+        return _error_response(404, f"There is no route {route} actived on this server")
+
+
+@app.get("/proxy_endpoint_url")
+async def proxy_endpoint_url(request: Request):
+    return {"Proxy URL": _parse_url_path_for_base_url(str(request.url))}
+
+
+@app.get("/disable_all_routes")
+async def disable_all_routes():
+    for route, status in active_routes.items():
+        if status.active:
+            active_routes[route].active = False
+    return {"status": "All routes disabled"}
+
+
+@app.get("/enable_all_routes")
+async def enable_all_routes():
+    for route, status in active_routes.items():
+        active_routes[route].active = True
+    return {"status": "All routes enabled"}
+
+
+@app.post("/shutdown_server")
+async def shutdown_server():
+    # This endpoint should be secured
+    global PID
+    try:
+        # sigterm
+        os.kill(PID, signal.SIGTERM)
+        sys.exit(0)
+    except ProcessLookupError:
+        return {"status": "Server is not running"}
+
+
+@app.middleware("http")
+async def rate_limit_middleware(request: Request, call_next: Callable):
```

Comment:
There are more sophisticated ways of handling rate limiting. I had a rough implementation using slowapi's Limiter() class, but in order to continue with that, the abstract function interface needs to be built so that its decorators can be applied to those dynamically created config-driven endpoint creation events. 
It's fairly complex, but doable.